IndexedDB/name-scopes.html

```diff
@@ -0,0 +1,151 @@
+        } else {
+          xKey = xKeyRoot;
+          yKey = yKeyRoot;
+        }
+        const path = `${databaseName}-${storeName}-${xKeyRoot}-${yKeyRoot}`;
+        store.put({ xKey: xKey, yKey: yKey, path: path });
+      }
+    }
+  }
+};
+
+// Creates two databases with identical structures.
+const buildDatabases = (testCase, useUniqueKeys) => {
+  return createNamedDatabase(testCase, 'x', database => {
+    return buildStores(database, 'x', useUniqueKeys);
+  }).then(database => {
+    return database.close();
+  }).then(() => createNamedDatabase(testCase, 'y', database => {
+    return buildStores(database, 'y', useUniqueKeys);
+  })).then(database => {
+    return database.close();
+  });
+};
+
+// Reads all the store's values using an index.
+//
+// Returns a Promise that resolves with an array of values.
+const readIndex = (testCase, index) => {
+  let resolve = null;
+  const promise = new Promise((_resolve, _reject) => resolve = _resolve);
+
+  const results = [];
+  const request = index.openCursor(IDBKeyRange.bound('a', 'z'), 'next');
+  request.onsuccess = () => {
+    const cursor = request.result;
+    if (cursor) {
+      results.push(cursor.value);
+      cursor.continue();
+    } else {
+      resolve(results);
+    }
+  }
+
+  return promise;
+}
+
+// Verifies that a database contains the expected records.
+const checkDatabaseContent =
+    (testCase, database, databaseName, usedUniqueKeys) => {
+  const promises = [];
+  const transaction = database.transaction(['x', 'y'], 'readonly');
+  for (let storeName of ['x', 'y']) {
+    const store = transaction.objectStore(storeName);
+    for (let indexName of ['x', 'y']) {
+      const index = store.index(indexName);
+
+      const promise = readIndex(testCase, index).then((results) => {
+        assert_array_equals(
+            results.map(result => `${result.path}:${result.pKey}`).sort(),
+            [`${databaseName}-${storeName}-x-x:1`,
+             `${databaseName}-${storeName}-x-y:2`,
+             `${databaseName}-${storeName}-y-x:3`,
+             `${databaseName}-${storeName}-y-y:4`],
+            'The results should include all records put into the store');
+
+        let expectedKeys = (usedUniqueKeys) ?
+            ['xx:xx', 'xy:yx', 'yx:xy', 'yy:yy'] : ['x:x', 'x:y', 'y:x', 'y:y'];
+        assert_array_equals(
+            results.map(result => `${result.xKey}:${result.yKey}`).sort(),
+            expectedKeys,
+            'The results should include all the index keys put in the store');
+
+        assert_array_equals(
+            results.map(result => result[`${indexName}Key`]),
+            results.map(result => result[`${indexName}Key`]).sort(),
+            'The results should be sorted by the index key');
+      });
+      promises.push(promise);
+    }
+  }
+
+  return Promise.all(promises);
+}
+
+promise_test(testCase => {
+  return buildDatabases(testCase, false).then(() => {
+    return openNamedDatabase(testCase, 'x', 1);
```

Comment:
I think this would more readable with some style changes:

* Don't use `x => { return y; }` when you can use just `x => y`
* Try line breaking before the `.then` 
* Avoid nested `then()`s if you can, i.e. just add more steps to the outer chain, if possible